packages/jsii/lib/assembler.ts

```diff
@@ -348,6 +492,31 @@ export class Assembler implements Emitter {
 
     if (!jsiiType) { return []; }
 
+    // Let's quickly verify the declaration does not collide with a submodule. Submodules get case-adjusted for each
+    // target language separately, so names cannot collide with case-variations.
+    for (const submodule of this._submodules) {
+      const candidates = Array.from(new Set([
+        submodule.name,
+        Case.camel(submodule.name),
+        Case.pascal(submodule.name),
+        Case.snake(submodule.name),
+      ]));
+      const colliding = candidates.find(name => `${this.projectInfo.name}.${name}` === jsiiType!.fqn);
+      if (colliding != null) {
+        const submoduleDecl = submodule.valueDeclaration ?? submodule.declarations[0];
+        this._diagnostic(node, ts.DiagnosticCategory.Error,
+          `Submodule "${submodule.name}" conflicts with "${jsiiType.name}". Restricted names are: ${candidates.join(', ')}`,
```

Comment:
I guess top level types should always be pascal case (interfaces, classes, enums), so I think by definition they can't conflict with namespace names...

Reply:
They can because the conflict space is `PascalCase` (preferred for C#),  `camelCase` (preferred for JS), `snake_case` (preferred for Python & Java).